File: Objective-C/Tests/QueryLegacyTest.m
//
//  QueryLegacyTest.m
//  CouchbaseLite
//
//  Copyright (c) 2026 Couchbase, Inc All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

#import "QueryTest.h"

/**
 Minimal smoke tests for the deprecated Query APIs:
 - CBLQueryDataSource +database: / +database:as:
 - CBLQueryExpression -isNullOrMissing / -notNullOrMissing
 - CBLQueryFullTextFunction +rank: / +matchWithIndexName:query:
 - CBLQueryFullTextExpression +indexWithName: / -match:

 The goal is only to touch each deprecated entry point and confirm its basic
 behavior, NOT to re-cover the full query test suites (which now use the
 collection-based / non-deprecated APIs).
 */
@interface QueryLegacyTest : QueryTest
@end

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"

@implementation QueryLegacyTest

// MARK: - CBLQueryDataSource +database: / +database:as:

- (void) testDataSourceDatabaseAndAlias {
    CBLMutableDocument* doc1 = [self createDocument: @"doc1"];
    [doc1 setValue: @"Scott" forKey: @"name"];
    [self saveDocument: doc1 collection: self.defaultCollection];

    CBLMutableDocument* doc2 = [self createDocument: @"doc2"];
    [doc2 setValue: @"Tiger" forKey: @"name"];
    [self saveDocument: doc2 collection: self.defaultCollection];

    // +database:
    CBLQuery* q = [CBLQueryBuilder select: @[kDOCID]
                                     from: [CBLQueryDataSource database: self.db]];
    uint64_t numRows = [self verifyQuery: q randomAccess: YES
                                    test: ^(uint64_t n, CBLQueryResult* r) { }];
    AssertEqual(numRows, 2u);

    // +database:as:
    CBLQuery* aliased = [CBLQueryBuilder select: @[kDOCID]
                                           from: [CBLQueryDataSource database: self.db as: @"main"]];
    numRows = [self verifyQuery: aliased randomAccess: YES
                           test: ^(uint64_t n, CBLQueryResult* r) { }];
    AssertEqual(numRows, 2u);
}

// MARK: - CBLQueryExpression -isNullOrMissing / -notNullOrMissing

- (void) testIsNullOrMissing {
    CBLMutableDocument* doc1 = [self createDocument: @"doc1"];
    [doc1 setValue: @"Scott" forKey: @"name"];
    [doc1 setValue: nil forKey: @"address"];
    [self saveDocument: doc1 collection: self.defaultCollection];

    CBLMutableDocument* doc2 = [self createDocument: @"doc2"];
    [doc2 setValue: @"Tiger" forKey: @"name"];
    [doc2 setValue: @"123 1st ave." forKey: @"address"];
    [self saveDocument: doc2 collection: self.defaultCollection];

    CBLQueryExpression* name = [CBLQueryExpression property: @"name"];
    CBLQueryExpression* address = [CBLQueryExpression property: @"address"];

    NSArray* tests = @[
        @[[name isNullOrMissing],     @0],
        @[[name notNullOrMissing],    @2],
        @[[address isNullOrMissing],  @1],
        @[[address notNullOrMissing], @1],
    ];

    for (NSArray* test in tests) {
        CBLQueryExpression* exp = test[0];
        NSUInteger expected = [test[1] unsignedIntegerValue];
        CBLQuery* q = [CBLQueryBuilder select: @[kDOCID]
                                         from: kDATA_SRC_DB
                                        where: exp];
        uint64_t numRows = [self verifyQuery: q randomAccess: YES
                                        test: ^(uint64_t n, CBLQueryResult* r) { }];
        AssertEqual((NSUInteger)numRows, expected, @"Failed case: %@", exp);
    }
}

// MARK: - CBLQueryFullTextFunction +rank: / +matchWithIndexName:query:

- (void) testFullTextFunctionDeprecatedRankAndMatch {
    [self loadJSONResource: @"sentences"];

    CBLQuerySelectResult* S_SENTENCE = [CBLQuerySelectResult property: @"sentence"];

    NSError* error;
    CBLFullTextIndex* index = [CBLIndexBuilder fullTextIndexWithItems: @[[CBLFullTextIndexItem property: @"sentence"]]];
    Assert([self.defaultCollection createIndex: index name: @"sentence" error: &error],
           @"Error when creating the index: %@", error);

    CBLQueryExpression* where = [CBLQueryFullTextFunction matchWithIndexName: @"sentence" query: @"'Dummie woman'"];
    CBLQueryOrdering* order = [[CBLQueryOrdering expression: [CBLQueryFullTextFunction rank: @"sentence"]]
                               descending];
    CBLQuery* q = [CBLQueryBuilder select: @[kDOCID, S_SENTENCE]
                                     from: kDATA_SRC_DB
                                    where: where
                                  orderBy: @[order]];
    uint64_t numRows = [self verifyQuery: q randomAccess: YES
                                    test: ^(uint64_t n, CBLQueryResult* r) { }];
    AssertEqual(numRows, 2u);
}

// MARK: - CBLQueryFullTextExpression +indexWithName: / -match:

- (void) testFullTextExpression {
    [self loadJSONResource: @"sentences"];

    CBLQueryFullTextExpression* SENTENCE = [CBLQueryFullTextExpression indexWithName: @"sentence"];
    CBLQuerySelectResult* S_SENTENCE = [CBLQuerySelectResult property: @"sentence"];

    NSError* error;
    CBLFullTextIndex* index = [CBLIndexBuilder fullTextIndexWithItems: @[[CBLFullTextIndexItem property: @"sentence"]]];
    Assert([self.defaultCollection createIndex: index name: @"sentence" error: &error],
           @"Error when creating the index: %@", error);

    CBLQueryExpression* where = [SENTENCE match: @"'Dummie woman'"];
    CBLQuery* q = [CBLQueryBuilder select: @[kDOCID, S_SENTENCE]
                                     from: kDATA_SRC_DB
                                    where: where];
    uint64_t numRows = [self verifyQuery: q randomAccess: YES
                                    test: ^(uint64_t n, CBLQueryResult* r) { }];
    AssertEqual(numRows, 2u);
}

// MARK: - CBLQuery -removeChangeListenerWithToken:

- (void) testRemoveQueryChangeListenerWithToken {
    CBLQuery* q = [CBLQueryBuilder select: @[kDOCID] from: kDATA_SRC_DB];

    XCTestExpectation* x = [self expectationWithDescription: @"query change"];
    x.assertForOverFulfill = NO;
    id<CBLListenerToken> token = [q addChangeListener: ^(CBLQueryChange* change) {
        AssertNil(change.error);
        [x fulfill];
    }];
    [self waitForExpectations: @[x] timeout: 5.0];

    // Remove using the deprecated API; should not crash.
    [q removeChangeListenerWithToken: token];
}

@end

#pragma clang diagnostic pop

File: Swift/Tests/QueryLegacyTest.swift
//
//  QueryLegacyTest.swift
//  CouchbaseLite
//
//  Copyright (c) 2026 Couchbase, Inc All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import XCTest
import CouchbaseLiteSwift

/// Minimal smoke tests for the deprecated Query APIs :
/// `DataSource.database(_:)`/`.as(_:)`, `Expression.isNullOrMissing()`/`notNullOrMissing()`,
/// `FullTextFunction.rank(_:)` (String) / `match(indexName:query:)`, `FullTextExpression`,
/// and the `SortOrder` typealias.
///
/// The goal is only to touch each deprecated entry point and confirm its basic behavior,
/// NOT to re-cover the full query test suites. The class is marked deprecated so calls to
/// the deprecated APIs below don't emit warnings.
@available(*, deprecated)
class QueryLegacyTest: CBLTestCase {

    // Local copy of QueryTest's helper (we subclass CBLTestCase, not the concrete QueryTest,
    // to avoid re-running the entire QueryTest suite via inheritance).
    let kDOCID = SelectResult.expression(Meta.id)

    // MARK: - DataSource.database(_:) / .as(_:)

    func testDataSourceDatabaseAndAlias() throws {
        let doc1 = createDocument("doc1")
        doc1.setValue("Scott", forKey: "name")
        try defaultCollection!.save(document: doc1)

        let doc2 = createDocument("doc2")
        doc2.setValue("Tiger", forKey: "name")
        try defaultCollection!.save(document: doc2)

        // DataSource.database(_:)
        let q = QueryBuilder.select(kDOCID).from(DataSource.database(db))
        var numRows = try verifyQuery(q) { (n, r) in }
        XCTAssertEqual(numRows, 2)

        // DataSource.database(_:).as(_:)
        let aliased = QueryBuilder.select(kDOCID).from(DataSource.database(db).as("main"))
        numRows = try verifyQuery(aliased) { (n, r) in }
        XCTAssertEqual(numRows, 2)
    }

    // MARK: - Expression.isNullOrMissing() / notNullOrMissing()

    func testIsNullOrMissing() throws {
        let doc1 = createDocument("doc1")
        doc1.setValue("Scott", forKey: "name")
        try defaultCollection!.save(document: doc1)

        let doc2 = createDocument("doc2")
        doc2.setValue("Tiger", forKey: "name")
        doc2.setValue("123 1st ave.", forKey: "address")
        try defaultCollection!.save(document: doc2)

        let name = Expression.property("name")
        let address = Expression.property("address")

        let tests: [(ExpressionProtocol, Int)] = [
            (name.isNullOrMissing(),     0),
            (name.notNullOrMissing(),    2),
            (address.isNullOrMissing(),  1),
            (address.notNullOrMissing(), 1),
        ]

        for (exp, expected) in tests {
            let q = QueryBuilder.select(kDOCID)
                .from(DataSource.collection(defaultCollection!))
                .where(exp)
            let numRows = try verifyQuery(q) { (n, r) in }
            XCTAssertEqual(Int(numRows), expected, "Failed case: \(exp)")
        }
    }

    // MARK: - FullTextFunction.rank(_:) (String) / match(indexName:query:)

    func testFullTextFunctionDeprecatedRankAndMatch() throws {
        try loadJSONResource(name: "sentences")

        let index = IndexBuilder.fullTextIndex(items: FullTextIndexItem.property("sentence"))
            .language(nil)
            .ignoreAccents(false)
        try defaultCollection!.createIndex(index, name: "sentence")

        let order = Ordering.expression(FullTextFunction.rank("sentence")).descending()
        let q = QueryBuilder
            .select([SelectResult.expression(Meta.id), SelectResult.property("sentence")])
            .from(DataSource.collection(defaultCollection!))
            .where(FullTextFunction.match(indexName: "sentence", query: "'Dummie woman'"))
            .orderBy([order])
        let numRows = try verifyQuery(q) { (n, r) in }
        XCTAssertEqual(numRows, 2)
    }

    // MARK: - FullTextExpression.index(_:).match(_:)

    func testFullTextExpression() throws {
        try loadJSONResource(name: "sentences")

        let index = IndexBuilder.fullTextIndex(items: FullTextIndexItem.property("sentence"))
            .language(nil)
            .ignoreAccents(false)
        try defaultCollection!.createIndex(index, name: "sentence")

        let sentence = FullTextExpression.index("sentence")
        let q = QueryBuilder
            .select([SelectResult.expression(Meta.id), SelectResult.property("sentence")])
            .from(DataSource.collection(defaultCollection!))
            .where(sentence.match("'Dummie woman'"))
        let numRows = try verifyQuery(q) { (n, r) in }
        XCTAssertEqual(numRows, 2)
    }

    // MARK: - SortOrder typealias

    func testSortOrderTypealias() throws {
        let doc1 = createDocument("doc1")
        doc1.setValue("Scott", forKey: "name")
        try defaultCollection!.save(document: doc1)

        let doc2 = createDocument("doc2")
        doc2.setValue("Tiger", forKey: "name")
        try defaultCollection!.save(document: doc2)

        // `SortOrder` is a deprecated public typealias for `QuerySortOrder`. It must be
        // module-qualified because it is ambiguous with Foundation.SortOrder.
        let order: CouchbaseLiteSwift.SortOrder = Ordering.property("name")
        let q = QueryBuilder.select(kDOCID)
            .from(DataSource.collection(defaultCollection!))
            .orderBy([order.ascending()])
        let numRows = try verifyQuery(q) { (n, r) in }
        XCTAssertEqual(numRows, 2)
    }

    // MARK: - Query.removeChangeListener(withToken:)

    func testRemoveQueryChangeListenerWithToken() throws {
        let q = QueryBuilder.select(kDOCID).from(DataSource.collection(defaultCollection!))

        let x = expectation(description: "query change")
        x.assertForOverFulfill = false
        let token = q.addChangeListener { (change) in
            XCTAssertNil(change.error)
            x.fulfill()
        }
        wait(for: [x], timeout: 5.0)

        // Remove using the deprecated API; should not crash.
        q.removeChangeListener(withToken: token)
    }
}
